gtk4/src/subclass/im_context.rs

```diff
@@ -0,0 +1,508 @@
+        *text_ptr = text.to_glib_none().0;
+        *cursor_index_ptr = cursor_idx;
+        return true.to_glib();
+    }
+    false.to_glib()
+}
+
+unsafe extern "C" fn im_context_preedit_changed<T: IMContextImpl>(ptr: *mut ffi::GtkIMContext) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    imp.preedit_changed(wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn im_context_preedit_end<T: IMContextImpl>(ptr: *mut ffi::GtkIMContext) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    imp.preedit_end(wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn im_context_preedit_start<T: IMContextImpl>(ptr: *mut ffi::GtkIMContext) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    imp.preedit_start(wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn im_context_reset<T: IMContextImpl>(ptr: *mut ffi::GtkIMContext) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    imp.reset(wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn im_context_retrieve_surrounding<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+) -> glib::ffi::gboolean {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    imp.retrieve_surrounding(wrap.unsafe_cast_ref()).to_glib()
+}
+
+unsafe extern "C" fn im_context_set_client_widget<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+    widgetptr: *mut ffi::GtkWidget,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+    let widget: Borrowed<Option<Widget>> = from_glib_borrow(widgetptr);
+
+    imp.set_client_widget(wrap.unsafe_cast_ref(), widget.as_ref().as_ref());
+}
+
+unsafe extern "C" fn im_context_set_cursor_location<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+    areaptr: *mut gdk::ffi::GdkRectangle,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+    let mut area = gdk::Rectangle::uninitialized();
+
+    imp.set_cursor_location(wrap.unsafe_cast_ref(), &mut area);
+    *areaptr = *area.to_glib_none().0;
+
+}
+
+unsafe extern "C" fn im_context_set_surrounding<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+    textptr: *const libc::c_char,
+    length: i32,
+    cursor_index: i32,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+    let text: GString = FromGlibContainer::from_glib_none_num(textptr, length as usize);
```

Comment:
The C code does uses `g_strndup (text, len);` to only get the passed length, does it make sense to a transfer: none here?

Reply:
What C code? And why not use `from_glib_borrow()` here?

Reply:
But then I would have to convert the gstring to bytes to only take `length` bytes & convert it back to a gstring?

Reply:
Pff... hmm.

```rust
let s = GString::from_glib_borrow(textptr);
f(&s[..length]);
```

or something like that maybe? The `from_glib_none()` there does a full copy of the string :)

Reply:
That worked fine :) thanks!